runtime/src/accounts_db.rs

```diff
@@ -1559,47 +1559,52 @@ impl AccountsDb {
                         let mut purges_in_root = Vec::new();
                         let mut purges = HashMap::new();
                         for pubkey in pubkeys {
-                            if let Some((locked_entry, index)) =
-                                self.accounts_index.get(pubkey, None, max_clean_root)
-                            {
-                                let slot_list = locked_entry.slot_list();
-                                let (slot, account_info) = &slot_list[index];
-                                if account_info.lamports == 0 {
-                                    purges.insert(
-                                        *pubkey,
-                                        self.accounts_index
-                                            .roots_and_ref_count(&locked_entry, max_clean_root),
-                                    );
-                                }
-
-                                // prune zero_lamport_pubkey set which should contain all 0-lamport
-                                // keys whether rooted or not. A 0-lamport update may become rooted
-                                // in the future.
-                                let has_zero_lamport_accounts = slot_list
-                                    .iter()
-                                    .any(|(_slot, account_info)| account_info.lamports == 0);
-                                if !has_zero_lamport_accounts {
-                                    self.accounts_index.remove_zero_lamport_key(pubkey);
-                                }
-
-                                // Release the lock
-                                let slot = *slot;
-                                drop(locked_entry);
-
-                                if self.accounts_index.is_uncleaned_root(slot) {
-                                    // Assertion enforced by `accounts_index.get()`, the latest slot
-                                    // will not be greater than the given `max_clean_root`
-                                    if let Some(max_clean_root) = max_clean_root {
-                                        assert!(slot <= max_clean_root);
+                            let remove_from_zero_lamports_list =
```

Comment:
Could we remove the redundant pubkeys instead on construction in `construct_candidate_clean_keys()` by using something like a `DashMap` instead of returning a vector with potential duplicates?

Reply:
Oh hehe I see this is solving a different problem. When reading the problem description I thought the duplicate lookup was for duplicate pubkeys, but it's for the extra `contains_key()` call below.

Reply:
I updated the description.